### PR TITLE
AuthServer identity providers support custom attribute / claim mapping for identity tokens

### PR DESCRIPTION
**Pull request by ddubson**
# Which other branches do you want a technical writer to cherry-pick this PR to (if any)?

It's best to PR to the most recent relevant branch and leave all the cherry-picking to the
writer, except where earlier branches require factual changes to the content.
For more information about the branches, see https://github.com/pivotal/docs-tap#branches

### CODE REVIEW

**Review by mamachanko — APPROVED** [review]
**[nit, non-blocking]** The use of
```
**`.some.field`**
```
might not help readability. I think that
```
`.some.field`
```
should be fine.

File: app-sso/reference/api/clientregistration.hbs.md [hunk 1]
```diff
@@ -57,6 +57,15 @@ Alternatively, you can interactively discover the spec with:
 kubectl explain clientregistrations.sso.apps.tanzu.vmware.com
 ```
 
+## <a id='scopes'></a> Scopes
+
+**Identity token**
+
+- **`openid`** scope must be included for identity tokens to be issued.
+- **`profile`** scope must be included for custom-mapped claims to be included in an issued identity token. e.g. `AuthServer.identityProviders[*].{openID,ldap,saml}.idToken.claims`. See [identity token claims mapping docs](../../service-operators/identity-providers.hbs.md#id-token-claims-mapping)
```
Comment (Kehrlann) [inline app-sso/reference/api/clientregistration.hbs.md:65]: **[suggestion]** as well as [standard openid claims](https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims), if available:  name, family_name, given_name, middle_name, nickname, 
preferred_username, profile, picture, website, gender, birthdate, zoneinfo, locale, and updated_at.

Also, **email**, **address** and **phone** are standard OIDC scopes that may be worth mentioning (particularly e-mail), see link above.

File: app-sso/service-operators/identity-providers.hbs.md [hunk 2]
```diff
@@ -662,21 +686,27 @@ spec:
             password: "password" # plain text
             roles:
               - "silly"
+            claims: # custom identity token claims
+              alt_address: "123 Fake Street"
+              middle_initial: "H"
```
Comment (Kehrlann) [inline app-sso/service-operators/identity-providers.hbs.md:691]: **[suggestion]** add standard OIDC profile / email claims, e.g.:

```yaml
claims:
  given_name: "Bert"
  family_name: "Muppet"
  email: "bert@muppets.example.com"
  middle_initial: "H"
  alt_address: "123 Sesame Street"
```

File: app-sso/service-operators/identity-providers.hbs.md [hunk 3]
```diff
@@ -690,6 +720,64 @@ There are multiple options for generating bcrypt hashes:
   htpasswd -bnBC 12 "" your-password-here | tr -d ':\n'
   ```
 
+## <a id="id-token-claims-mapping"></a> Identity token claims mapping
+
+> **Note** This section is applicable to OpenID, LDAP, and SAML (experimental) identity provider configurations.
+
+`AuthServer` identity provider specification allows for configuring custom identity token claim mappings.
```
Comment (Kehrlann) [inline app-sso/service-operators/identity-providers.hbs.md:727]: **[suggestion]**

> Service Operators can control which claims appear in an `AuthServer`-issued `id_token`, and how to obtain this value from an upstream identity provider.

**[suggestion]** add an example

> For example, consider the following configuration:
> ```yaml
> spec:
>   identityProviders:
>   - name: my-identity-provider
>     openid:
>       idToken:
>         claims:
>           - fromUpstream: "title"
>             toClaim: "job_tile"
>```
> With this, if the upstream OpenID identity provider makes a "title" claim available with value "developer", then the corresponding `AuthServer`-issued `id_token` will contain a claim `"job_title": "developer"`.

File: app-sso/service-operators/identity-providers.hbs.md [hunk 4]
```diff
@@ -690,6 +720,64 @@ There are multiple options for generating bcrypt hashes:
...
+
+This field also allows for mapping [standard OpenID claims](https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims) as well (
+such as `given_name`, `family_name`, etc.). The field types will be preserved as per original OpenID
+specification. For example, the `given_name` claim is defined as a string in the OpenID specification, and so if a
+custom claim is mapped into this field, the value of the custom claim will be coerced into a string type -- if the custom
+claim is an array type, the first value of the array will be used.
```
Comment (Kehrlann) [inline app-sso/service-operators/identity-providers.hbs.md:733]: **[suggestion]** Reword a bit:

> This field also allows for mapping [standard OpenID claims](https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims) (such as `given_name`, `family_name`, etc.) as well as non-standard claim (such as the `job_title` claim in the example above).
> When mapping standard claims, the field types will be preserved [...]

File: app-sso/service-operators/identity-providers.hbs.md [hunk 5]
```diff
@@ -690,6 +720,64 @@ There are multiple options for generating bcrypt hashes:
...
+
+> **Important** To be able to use custom claims defined within `.spec.identityProviders[*].{openid,ldap,saml}.idToken.claims`,
+> you **must** include the `profile` scope within your `ClientRegistration` resource.
```
Comment (Kehrlann) [inline app-sso/service-operators/identity-providers.hbs.md:736]: **[suggestion]**

```suggestion
> **Important** For custom claims defined `.spec.identityProviders[*].{openid,ldap,saml}.idToken.claims` to be available in `AuthServer`-issued `id_token`s,
> App Operators **must** include the `profile` scope within their `ClientRegistration` resource.
```

**Review by Kehrlann — APPROVED** [review]
A few suggestions but nothing blocking.

File: app-sso/reference/api/clientregistration.hbs.md (hunk 1 above)
Comment (ddubson) [reply app-sso/reference/api/clientregistration.hbs.md:65]: Done

File: app-sso/service-operators/identity-providers.hbs.md (hunk 2 above)
Comment (ddubson) [reply app-sso/service-operators/identity-providers.hbs.md:691]: Done

File: app-sso/service-operators/identity-providers.hbs.md (hunk 3 above)
Comment (ddubson) [reply app-sso/service-operators/identity-providers.hbs.md:727]: Good call, done.

File: app-sso/service-operators/identity-providers.hbs.md (hunk 4 above)
Comment (ddubson) [reply app-sso/service-operators/identity-providers.hbs.md:733]: Done

File: app-sso/service-operators/identity-providers.hbs.md (hunk 5 above)
Comment (ddubson) [reply app-sso/service-operators/identity-providers.hbs.md:736]: Done.

File: app-sso/service-operators/identity-providers.hbs.md [hunk 6]
```diff
@@ -690,6 +720,64 @@ There are multiple options for generating bcrypt hashes:
...
+          - fromUpstream: "" # identity provider claim name
+            toClaim: ""      # claim that will be part of an AppSSO id_token
+```
+
+### Constraints
+
+The following are identity token mapping constraints that must be followed:
+
+- Multiple upstream identity provider claims cannot be mapped to a single `AuthServer` identity token claim.
+- LDAP upstream identity provider claims are case-insensitive, while OpenID and SAML claims are case-sensitive.
```
[the remark this answers is not in the document]
Comment (ddubson) [reply app-sso/service-operators/identity-providers.hbs.md:759]: Done

File: app-sso/service-operators/identity-providers.hbs.md [hunk 7]
```diff
@@ -690,6 +720,64 @@ There are multiple options for generating bcrypt hashes:
...
+- `AuthServer` identity token claims are case-sensitive.
```
[the remark this answers is not in the document]
Comment (ddubson) [reply app-sso/service-operators/identity-providers.hbs.md:760]: Done.

**Comment by ddubson** [comment]
Opening TAP docs PR. Closing this PR.

**Pull request closed without merging**